### DISCUSSION TITLE
Add a new checkbox to prompt the user for confirmation before saving and overwriting or loading any slot or global state

### DISCUSSION THREAD

**Discussion by ghost** [grumpycoders/pcsx-redux#1090] [Ideas]
Right now the user is **NOT** prompted for confirmation and he or she might **accidentally** and **mistakenly** overwrite his or her progress when trying to either saving or loading any state to or from non-empty slot or global.

**Comment by wheremyfoodat** [comment]
This is by design and is most likely not feasible to change. The main idea is that accidentally loading/saving a savestate from the menu bar is almost impossible so not much point in adding a confirmation menu cause it would just get in the way.

For the shortcut buttons, they're typically used in occasions where timing and fast reflexes are important (eg when you're trying to make a savestate before an enemy in a game bonks you). Adding a confirmation in this case would screw these people (which is frankly the majority) pretty hard.

**Comment by ghost** [comment]
> This is by design and is most likely not feasible to change. The main idea is that accidentally loading/saving a savestate from the menu bar is almost impossible so not much point in adding a confirmation menu cause it would just get in the way.
> 
> For the shortcut buttons, they're typically used in occasions where timing and fast reflexes are important (eg when you're trying to make a savestate before an enemy in a game bonks you). Adding a confirmation in this case would screw these people (which is frankly the majority) pretty hard.

This is a checkbox in the GUI, not hardcoded behind the scenes, and the prompt for confirmation is shown only when the checkbox is checked.

In occasions where timing and fast reflexes are important the user may want to leave this checkbox unchecked to not be prompted for confirmation every time saving or loading a state but otherwise the user may want to leave it checked if saving and loading states is not so frequent action.

One example is when the user edited the memory of the game to enable god mode and he or she can't lose in the game as a result.

In this case the user doesn't have to save state before an enemy attack and load state if he or she failed to avoid that enemy attack because he or she is invulnerable to all enemy attacks and doesn't have to avoid any of them at all.

In this case the user may want to save state when he or she can no longer continue playing, because he or she needs to go to work or eat lunch or go shower or go to sleep and etc.

Later he or she will load state to continue playing where he or she left off. In this case save and load state is not so frequent actions and the user may want to make sure that he or she doesn't do any mistake when he or she goes to save or load a state.

In this case the prompt for confirmation may save him or her if for example he or she wants to save his or her progress but his or her hand **slipped** and accidentally clicked on the "Load global state" instead of the "Save global state" and lost a few hours of progress as a result!

PCSX2 1.6.0 never asks the user if he or she is sure that he or she wants to overwrite an existing saved state file when saving state to a named file.

Because of this it has already happened to me that I both accidentally and mistakenly overwritten a saved state file that I didn't want to overwrite and as a result I lost all the progress of countless hours that was saved in that state file!

If PCSX2 1.6.0 would ask me before overwrite I could keep this progress saved by just clicking on the No button to not overwrite the saved state file and lose all the progress was saved in that state file.

I don't know if this problem was solved in the most recent nightly build and release of PCSX2 though.

I have already thought about all possible scenarios and outcomes and my conclusion was to propose to make it a checkbox in the GUI rather than hardcoded of course. Every single user will decide for his or her likings if he or she wants this feature enabled or not for his or her convenience.

Of course that making it hardcoded is a **terrible** idea and do NOT do this because of what you said!

In general hardcoding is bad practice and should be avoided where possible!

Think how awful it would have been if "Use raw input for mouse" was hardcoded and always enabled and the user couldn't disable it at all?

This is exactly what checkboxes are for after all.

And what do you mean by shortcut buttons exactly? Are you saying that this is possible to quickly save and load states by just pressing two keys on the keyboard like quickly pause with F6 and quickly resume with F5?

If yes then there is no logic to prompt the user for confirmation when he or she wants to quickly save or load state with a key press. Just make sure that the key for save state is not too close to the key for load state and this is alright.

In this case there won't be any prompt for confirmation even if the checkbox is checked.

The user will be prompted for confirmation only when the user saves or loads state with a mouse button click through the File menu and the checkbox to show this prompt at all is checked of course.

Also accidentally saving a global state or loading a global state with a mouse button click through the File menu is more likely to happen than what you think because both "Save global state" and "Load global state" are next to each other in the File menu and they both appear small in the screen.

But accidentally saving a state to a slot or loading a state from a slot with a mouse button click on the other hand is pretty unlikely according to their menu design though and in this case you are right even though this still can happen that the user only thinks which slot he or she wants to choose but he or she doesn't notice if he or she chose "Save state slots" or "Load state slots" and he or she may perform the wrong action that may result in losing a lot of progress. This already happened to me in the distant past when I used an old build and release of PCSX2 today, which was the newest in that time, when you could only save and load state but only to and from a slot and still not to and from a named file.

Perhaps my concern which caused me to start this feature request from the beginning was the fear to accidentally click the "Save global state" or the "Load global state" in the File menu and I didn't think about the both save and load state slots which I can use instead for safety.

Perhaps this was a bad idea to add both "Save global state" and "Load global state" to the File menu without any prompt for confirmation on click in the first place because if both "Save global state" and "Load global state" were not exist in the File menu but only save and load state slots then I wouldn't start this feature request from the beginning!

Also why both "Save global state" and "Load global state" exist in the File menu at all while the user can use state slots to save his or her progress and load to continue later?

If there is a reason to add both "Save global state" and "Load global state" to the File menu then at least show this prompt for confirmation but only when either "Save global state" or "Load global state" in the File menu was clicked with the mouse.

In any other way to save or load state (not global), like using slots or pressing a key on the keyboard, this prompt for confirmation will screw his or her users like you said and as a result the prompt won't show and the save or load state will be just done immediately if not instantly.

> Are you sure that you want to save to global state?

Will be shown to the user but only when the "Save global state" in the File menu was clicked with the mouse.

> Are you sure that you want to load from global state?

Will be shown to the user but only when the "Load global state" in the File menu was clicked with the mouse.

Both save and load states with slots and key presses no need confirmation of course and will be done immediately if not instantly to not screw their users.

The new checkbox that I proposed and suggested to add at the beginning is no longer needed on second thought.

So what do you think about my new idea, proposal or suggestion?

I will be happy to know your opinions!

Perhaps I should start a new discussion about this new idea!

Okay I have started this discussion [here](https://github.com/grumpycoders/pcsx-redux/discussions/1119)

I have also started another discussion about quick load state [here](https://github.com/grumpycoders/pcsx-redux/discussions/1147)
